File: python/sorting.py
import random
import sys
import time

def bubbleSort(arr):
	n = len(arr)

	for i in range(n):

		for j in range(0, n-i-1):

			if arr[j] > arr[j + 1]:
				arr[j], arr[j + 1] = arr[j + 1], arr[j]


def selectionSort(arr):
	for i in range(len(arr)):

		min_idx = i
		for j in range(i + 1, len(arr)):
			if arr[min_idx] > arr[j]:
				min_idx = j

		arr[i], arr[min_idx] = arr[min_idx], arr[i]


def mergeSort(arr, l, r):
	if l < r:
		m = (l + (r - 1)) / 2

		mergeSort(arr, l, m)
		mergeSort(arr, m + 1, r)
		merge(arr, l, m, r)


def merge(arr, l, m, r):
	n1 = m - l + 1
	n2 = r - m

	L = [0] * (n1)
	R = [0] * (n2)

	for i in range(0 , n1):
		L[i] = arr[l + i]

	for j in range(0 , n2):
		R[j] = arr[m + 1 + j]

	i = 0
	j = 0
	k = l

	while i < n1 and j < n2 :
		if L[i] <= R[j]:
			arr[k] = L[i]
			i += 1
		else:
			arr[k] = R[j]
			j += 1
		k += 1

	while i < n1:
		arr[k] = L[i]
		i += 1
		k += 1

	while j < n2:
		arr[k] = R[j]
		j += 1
		k += 1

def partition(arr, low, high):
	i = (low - 1)
	pivot = arr[high]

	for j in range(low, high):
		if arr[j] <= pivot:
			i += 1
			arr[i], arr[j] = arr[j], arr[i]

	arr[i + 1], arr[high] = arr[high], arr[i + 1]
	return (i + 1)

def quickSort(arr, low, high):
	if low < high:
		pi = partition(arr, low, high)

		quickSort(arr, low, pi - 1)
		quickSort(arr, pi + 1, high)

if len(sys.argv) != 2:
	print("usage: python sorting.py <size>")
	sys.exit()
size = int(sys.argv[1])

randomArray = [0] * size
for i in range(size):
	randomArray[i] = i
random.shuffle(randomArray)

# run and time mergesort
m = list(randomArray)
start_time = time.time()
mergeSort(m, 0, len(m) - 1)
m_time = time.time() - start_time

# run and time quicksort
q = list(randomArray)
start_time = time.time()
quickSort(q, 0, len(q) - 1)
q_time = time.time() - start_time

# run and time selection sort
s = list(randomArray)
start_time = time.time()
selectionSort(s)
s_time = time.time() - start_time

# run and time bubble sort
b = list(randomArray)
start_time = time.time()
bubbleSort(b)
b_time = time.time() - start_time

print("~~~~~~~~~~RESULTS~~~~~~~~~~")
print("<Algorithm>       <Runtime>")
print("MergeSort:        %s" %m_time)
print("QuickSort:        %s" %q_time)
print("SelectionSort:    %s" %s_time)
print("BubbleSort:       %s" %b_time)

File: c/sorting.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>

void swap( int *a, int *b )
{
  int temp = *a;
  *a = *b;
  *b = temp;
}

void shuffle( int arr[], int n )
{
  srand( time( NULL ) );

  for ( int i = n - 1; i > 0; i-- ) {
    int j = rand() % ( i + 1 );
    swap( &arr[ i ], &arr[ j ] );
  }
}

void bubbleSort( int arr[], int n )
{
  int i, j;
  for ( i = 0; i < n - 1; i++ ) {
    for ( j = 0; j < n - i - 1; j++ ) {
      if ( arr[ j ] > arr[ j + 1 ] ) {
        swap( &arr[ j ], &arr[ j + 1 ] );
      }
    }
  }
}

void selectionSort( int arr[], int n )
{
  int i, j, minIdx;

  for ( i = 0; i < n - 1; i++ ) {
    minIdx = i;
    for ( j = i + 1; j < n; j++ ) {
      if ( arr[ j ] < arr[ minIdx ] ) {
        minIdx = j;
      }
    }
    swap( &arr[ minIdx ], &arr[ i ] );
  }
}

void merge( int arr[], int l, int m, int r )
{
  int i, j, k;
  int n1 = m - l + 1;
  int n2 =  r - m;

  int L[ n1 ], R[ n2 ];

  for (i = 0; i < n1; i++) {
      L[ i ] = arr[ l + i ];
  }

  for ( j = 0; j < n2; j++ ) {
      R[ j ] = arr[ m + 1+ j ];
  }

  i = 0;
  j = 0;
  k = l;
  while ( i < n1 && j < n2 ) {
    if ( L[ i ] <= R[ j ] ) {
      arr[ k ] = L[ i ] ;
      i++;
    } else {
      arr[ k ] = R[ j ];
      j++;
    }
    k++;
  }

  while ( i < n1 ) {
    arr[ k ] = L[ i ];
    i++;
    k++;
  }

  while ( j < n2 ) {
    arr[ k ] = R[ j ];
    j++;
    k++;
  }
}

void mergeSort( int arr[], int l, int r )
{
  if ( l < r ) {
    int m = l + ( r - l ) / 2;

    mergeSort( arr, l, m );
    mergeSort( arr, m+1, r );

    merge( arr, l, m, r );
  }
}

int partition( int arr[], int low, int high )
{
  int pivot = arr[ high ];
  int i = ( low - 1 );

  for ( int j = low; j <= high- 1; j++ ) {
    if ( arr[ j ] <= pivot ) {
      i++;
      swap( &arr[ i ], &arr[ j ] );
    }
  }
  swap( &arr[ i + 1 ], &arr[ high ] );
  return ( i + 1 );
}

void quickSort( int arr[], int low, int high )
{
  if ( low < high ) {
    int pi = partition( arr, low, high );

    quickSort( arr, low, pi - 1 );
    quickSort( arr, pi + 1, high );
  }
}

void printArray( int arr[], int n )
{
  for ( int i = 0; i < n; i++ ) {
    printf( "%d ", arr[ i ] );
  }
  printf( "\n" );
}

void printResults( double b, double s, double m, double q )
{
  printf( "~~~~~~~~~~RESULTS~~~~~~~~~~\n" );
  printf( "<Algorithm>       <Runtime>\n" );
  printf( "MergeSort         %lf\n", m );
  printf( "QuickSort         %lf\n", q );
  printf( "SelectionSort     %lf\n", s );
  printf( "BubbleSort        %lf\n", b );
}

int main( int argc, char *argv[] )
{
  if ( argc != 2 ) {
    printf( "usage: ./sorting <size>\n" );
    exit( 1 );
  }

  // convert cmd arg to integer
  int size = atoi( argv[ 1 ] );

  // generate random array
  int randomArray[ size ];
  for ( int i = 0; i < size; i++ ) {
    randomArray[ i ] = i;
  }
  shuffle( randomArray, size );

  // execute various sorting methods
  clock_t t;

  // bubble sort
  int b[ size ];
  memcpy( b, randomArray, sizeof( b ) );

  t = clock();
  bubbleSort( b, size );
  t = clock() - t;

  double bTime = ( ( double ) t ) / CLOCKS_PER_SEC;

  // selection sort
  int s[ size ];
  memcpy( s, randomArray, sizeof( s ) );

  t = clock();
  selectionSort( s, size );
  t = clock() - t;

  double sTime = ( ( double ) t ) / CLOCKS_PER_SEC;

  // merge sort
  int m[ size ];
  memcpy( m, randomArray, sizeof( m ) );

  t = clock();
  mergeSort( m, 0, size - 1 );
  t = clock() - t;

  double mTime = ( ( double ) t ) / CLOCKS_PER_SEC;


  // quick sort
  int q[ size ];
  memcpy( q, randomArray, sizeof( q ) );

  t = clock();
  quickSort( q, 0, size - 1 );

  t = clock() - t;
  double qTime = ( ( double ) t ) / CLOCKS_PER_SEC;

  printResults( bTime, sTime, mTime, qTime );
}
